components/script/script_module.rs

```diff
@@ -657,6 +639,161 @@ impl ModuleTree {
+    fn fetch_module_descendants(
+        &self,
+        owner: &ModuleOwner,
+        destination: Destination,
+        credentials_mode: CredentialsMode,
+        parent_identity: ModuleIdentity,
+    ) {
+        debug!("Start to load dependencies of {}", self.url.clone());
+
+        let global = owner.global();
+
+        self.set_status(ModuleStatus::FetchingDescendants);
+
+        match self.resolve_requested_modules(&global) {
+            // Step 3.
+            Ok(requested_urls) if requested_urls.len() == 0 => {
+                self.advance_finished_and_link(&global);
+            },
+            Ok(mut requested_urls) => {
+                self.descendant_urls
+                    .borrow_mut()
+                    .extend(requested_urls.clone());
+
+                let parent_identities = self.parent_identities.borrow();
+
+                for parent_identity in parent_identities.iter() {
+                    match parent_identity {
+                        ModuleIdentity::ScriptId(_) => continue,
+                        ModuleIdentity::ModuleUrl(parent_url) => {
+                            if requested_urls.contains(&parent_url.clone()) {
+                                self.advance_finished_and_link(&global);
+                                requested_urls.remove(&parent_url.clone());
+                            }
+                        },
+                    }
+                }
+
+                for requested_url in requested_urls {
+                    // https://html.spec.whatwg.org/multipage/#internal-module-script-graph-fetching-procedure
+                    let visited_urls = {
+                        let visited_urls = self.visited_urls.borrow().clone();
+
+                        // Step 1.
+                        assert!(visited_urls.get(&requested_url).is_some());
+                        visited_urls
+                    };
+
+                    // Step 2.
+                    fetch_single_module_script(
+                        owner.clone(),
+                        requested_url.clone(),
+                        visited_urls,
+                        destination.clone(),
+                        Referrer::Client,
+                        ParserMetadata::NotParserInserted,
+                        "".to_owned(), // integrity
+                        credentials_mode.clone(),
+                        Some(parent_identity.clone()),
+                        false,
+                    );
+                }
+            },
+            Err(error) => {
+                self.set_error(Some(error));
+                self.advance_finished_and_link(&global);
+            },
+        }
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#fetch-the-descendants-of-and-link-a-module-script
+    /// step 4-7.
+    fn advance_finished_and_link(&self, global: &GlobalScope) {
+        // FIXME(cybai): try to find a better way for checking `ready` descendants
```

Comment:
@Manishearth thanks for your review first!

I found most your comments are about confusion in this `advance_finished_and_link` and that's because we currently save a set of parent identities and descendant sets.

So, when we're trying to finish a module, we will need to `notify` its parent (the parent for loop) or `notify` its descendants (when it's a cyclic module) to finish.

Instead of doing so many loops, I think I'd like to try doing the chromium way. They make a Module [hold a counter](https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/loader/modulescript/module_tree_linker.h;l=120) and increment when fetching more descendants and decrement when the descendants fetched.

I think using this way should make this function less scary. We might still need to loop parents to notify them to finish but at least it should be easier to understand and no need to do recursive steps. (also, we can remove this FIXME)

Please let me do more works on this 🙏 

Will try to ping you for review again when it's ready! Thanks!

Reply:
This seems good! Let me know when ready!